dali/pipeline/operators/color/brightness_contrast.h

```diff
@@ -0,0 +1,193 @@
+}
+
+
+template <>
+void assign_argument_value<CPUBackend>(const OpSpec &spec, const std::string &arg_name,
+                                       argument_t<CPUBackend> &arg) {
+  std::vector<float> tmp;
+  GetSingleOrRepeatedArg(spec, tmp, arg_name);
+  arg = tmp[0];
+}
+
+
+template <>
+void assign_argument_value<GPUBackend>(const OpSpec &spec, const std::string &arg_name,
+                                       argument_t<GPUBackend> &arg) {
+  GetSingleOrRepeatedArg(spec, arg, arg_name);
+}
+
+}  // namespace detail
+
+
+template <typename Backend>
+class BrightnessContrast : public Operator<Backend> {
+ public:
+  explicit BrightnessContrast(const OpSpec &spec) :
+          Operator<Backend>(spec),
+          output_type_(spec.GetArgument<DALIDataType>(detail::kOutputType)) {
+    detail::assign_argument_value<Backend>(spec, detail::kBrightness, brightness_);
+    detail::assign_argument_value<Backend>(spec, detail::kContrast, contrast_);
+    if (std::is_same<Backend, GPUBackend>::value) {
+      kernel_manager_.Resize(1, 1);
+    } else {
+      kernel_manager_.Resize(num_threads_, batch_size_);
+    }
+  }
+
+
+  ~BrightnessContrast() = default;
+
+
+  DISABLE_COPY_MOVE_ASSIGN(BrightnessContrast);
+
+
+ protected:
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+
+  bool SetupImpl(std::vector<::dali::OutputDesc> &output_desc,
+                 const ::dali::workspace_t<Backend> &ws) override {
+    const auto &input = ws.template InputRef<Backend>(0);
+    const auto &output = ws.template OutputRef<Backend>(0);
+    output_desc.resize(1);
+
+    TYPE_SWITCH(output_type_, type2id, OutputType, (uint8_t, int16_t, int32_t, float), (
+        {
+          TypeInfo type;
+          type.SetType<OutputType>(output_type_);
+          output_desc[0] = {input.shape(), type};
+        }
+    ), DALI_FAIL("Unsupported image type"))  // NOLINT
+    return true;
+  }
+
+  /**
+   * So that compiler wouldn't complain, that
+   * "overloaded virtual function `dali::Operator<dali::CPUBackend>::RunImpl` is only partially
+   * overridden in class `dali::brightness_contrast::BrightnessContrast<dali::CPUBackend>`"
+   */
+  using Operator<Backend>::RunImpl;
+
+
+  void RunImpl(Workspace<Backend> &ws) {
+    const auto &input = ws.template Input<Backend>(0);
+    auto &output = ws.template Output<Backend>(0);
+
+    TYPE_SWITCH(input.type().id(), type2id, InputType, (uint8_t, int16_t, int32_t, float), (
+        TYPE_SWITCH(output_type_, type2id, OutputType, (uint8_t, int16_t, int32_t, float), (
+        {
+            using BrightnessContrastKernel =
+                    detail::BrightnessContrastKernel<Backend, OutputType, InputType, 3>;
+            auto tvin = view<const InputType, 3>(input);
+            auto tvout = view<OutputType, 3>(output);
+            kernel_manager_.Initialize<BrightnessContrastKernel>();
+            kernels::KernelContext ctx;
+            kernel_manager_.Setup<BrightnessContrastKernel>(ws.data_idx(),
```

Comment:
What's your opinion on calling `kernel::Setup` in Operator::SetupImpl`?

Reply:
Generally speaking, I'd say that it should be this way. However, here we have a special case. Meaning, it's not necessary to call `Setup` in `SetupImpl`. On the other hand, implementing it this way would require major change with duplicating `TYPE_SWITCH`es, which obfuscates the code. While it would be possible to do this implementation without such duplication, the abstraction layer for this is not ready yet and will require some time to evolve. Keeping all this in mind, plus taking into account, that this PR is already quite a time here (I already rebased it twice), I'd like to merge it now and, when it proves to be necessary, add the abstraction layer later.